test/io/test_io.py

```diff
@@ -1063,3 +1063,33 @@ def test_succeed_w_role_having_superuser_settings(defaultenv):
         response = postgrest.session.get("/projects", headers=headers)
         print(response.text)
         assert response.status_code == 200
+
+
+def test_fail_with_invalid_dbname_and_automatic_recovery_disabled(defaultenv):
+    "Should fail without retries when automatic recovery is disabled and dbname is invalid"
+    dbname = "INVALID"
+    uri = f'postgresql://?dbname={dbname}&host={defaultenv["PGHOST"]}&user={defaultenv["PGUSER"]}'
+    env = {
+        **defaultenv,
+        "PGRST_DB_URI": uri,
+        "PGRST_DB_POOL_AUTOMATIC_RECOVERY": "false",
+    }
+
+    with run(env=env, wait_for_readiness=False) as postgrest:
+        exitCode = wait_until_exit(postgrest)
+        assert exitCode == 1
+
+
+def test_fail_with_automatic_recovery_disabled_and_terminated_using_query(defaultenv):
+    "Should fail without retries when automatic recovery is disabled and pg_terminate_backend(pid) is called"
+
+    env = {
+        **defaultenv,
+        "PGRST_DB_POOL_AUTOMATIC_RECOVERY": "false",
+    }
+
+    try:
+        with run(env=env) as postgrest:
+            postgrest.session.get("/rpc/terminate_pgrst")
+    except ConnectionError:
+        return True
```

Comment:
This test does not fail on the main branch when it should. I think verifying if postgrest stopped running is best:

```suggestion
    with run(env=env) as postgrest:
        postgrest.session.post("/rpc/terminate_pgrst")

        exitCode = wait_until_exit(postgrest)
        assert exitCode == 1
```

Still, doing the above does not close the connection and the test fails. It gives a  "Database client error. Retrying the connection." message and reloads PostgREST afterwards.  This *could* be considered "Automatic recovery", so maybe this shouldn't be happening when `db-pool-automatic-recovery=true`? If so, I think the thread must die here too:

https://github.com/PostgREST/postgrest/blob/531a183b44b36614224fda432335cdaa356b4a0a/src/PostgREST/App.hs#L122-L125

Reply:
> This _could_ be considered "Automatic recovery", so maybe this shouldn't be happening when `db-pool-automatic-recovery=true`? If so, I think the thread must die here too:

@laurenceisla I think you mean when `db-pool-automatic-recovery=false`. Yes, this shouldn't be happening. However, testing locally, the `terminate_pgrst()` terminates the postgrest without retries. Maybe there is something specific related to IO tests that we are missing?

Reply:
> select terminate_pgrst();

Yes, this works for me too. This is what I meant when I wrote "testing locally".

> Does using the curl command stop PostgREST on your end?

No. It reloads the connection.

Reply:
> No. It reloads the connection.

Yes, this happens with IO tests too, because they do the same request to PostgREST and reload the connection after that.

You're right that the IO tests are missing something, in this case a direct PostgreSQL execution. Luckily Python allows executing commands directly using `os.sytem("command")`. Then we can use `psql` to execute the SQL command like this:

```python
os.system(f'psql -d {defaultenv["PGDATABASE"]} -U {defaultenv["PGUSER"]} -h {defaultenv["PGHOST"]} --set ON_ERROR_STOP=1 -a -c "SELECT terminate_pgrst()"')
```

I replaced this in my original suggestion. This would solve the reloading when closing the connections outside PostgREST.

---

Now, there's the debate with reloading while doing a request. If we consider that it should not reload then the change must be done where I specified earlier.

Reply:
> ```os.system(f'psql -d {defaultenv["PGDATABASE"]} -U {defaultenv["PGUSER"]} -h {defaultenv["PGHOST"]} --set ON_ERROR_STOP=1 -a -c "SELECT terminate_pgrst()"')```

The above worked smoothly.

```haskell
if configDbPoolAutomaticRecovery appConf
  then when (Response.isServiceUnavailable response) connWorker
else when (Response.isServiceUnavailable response) (killThread $ AppState.getMainThreadId appState)
```
For the `rpc` call, I tried above as you suggested earlier, however this didn't work either. It throws a `ConnectionError` exception and then reloads PostgREST.

Reply:
> For the rpc call, I tried above as you suggested earlier, however this didn't work either. It throws a ConnectionError exception and then reloads PostgREST.

Hmmm... Yes, now that I see it, the request is not finished there, thus the error. The try/catch that you did before may be needed here. 

However, now that I see it more carefully, I think we could leave it as is (not modifying `App.hs`), because what this PR is looking for is to stop PostgREST when the DB connection is closed from outside. If the connection is really lost, the `connWorker` will fail with the `NotConnected` status anyways, so this extra verification is not needed here. In summary, closing PostgreSQL from within PostgREST is a special case that is not contemplated in the issue.